### PR TITLE
Handle NPE when healthz EP is invoked before server startup

### PR DESCRIPTION
**Pull request by imsuneth**
Fix for https://github.com/wso2/api-manager/issues/164
Now the NPE is handled and further the connection is shutdown when a request is made before the server getting started.

### CODE REVIEW

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java [hunk 1]
```diff
@@ -46,32 +49,59 @@ public MultiListenerServerIODispatch(final Map<Integer, NHttpServerEventHandler>
     protected void onConnected(final DefaultNHttpServerConnection defaultNHttpServerConnection) {
         int localPort = defaultNHttpServerConnection.getLocalPort();
         NHttpServerEventHandler handler = handlers.get(localPort);
-        try {
-            handler.connected(defaultNHttpServerConnection);
-        } catch (final Exception ex) {
-            handler.exception(defaultNHttpServerConnection, ex);
+        if (handler == null) {
+            log.error("Listener has not started yet on " + localPort + ". Hence shutting down the connection");
```
Comment (sajinieKavindya) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java:53]: Shall we change the error log to something like, "The request to the port + localPort + cannot be processed further as the Listener has not yet started on port" + localPort?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java [hunk 2]
```diff
@@ -46,32 +49,59 @@ public MultiListenerServerIODispatch(final Map<Integer, NHttpServerEventHandler>
...
+            try {
+                defaultNHttpServerConnection.shutdown();
+            } catch (IOException ex) {
+                log.error("Error shutting down the connection", ex);
```
Comment (sajinieKavindya) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java:57]: without logging an error message, shall we simply ignore the exception as in https://github.com/wso2/wso2-synapse/blob/d66dd9a13362d4b998378779391d22f6223b9088/modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/connections/SourceConnections.java#L157

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java (hunk 1 above)
Comment (imsuneth) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java]: Done

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java (hunk 2 above)
Comment (imsuneth) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java]: Done

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java [hunk 3]
```diff
@@ -46,32 +49,59 @@ public MultiListenerServerIODispatch(final Map<Integer, NHttpServerEventHandler>
...
 
     @Override
     protected void onClosed(final DefaultNHttpServerConnection defaultNHttpServerConnection) {
         int localPort = defaultNHttpServerConnection.getLocalPort();
         NHttpServerEventHandler handler = handlers.get(localPort);
-        try {
-            handler.closed(defaultNHttpServerConnection);
-        } catch (final Exception ex) {
-            handler.exception(defaultNHttpServerConnection, ex);
+        if (handler == null) {
```
[the remark this answers is not in the document]
Comment (imsuneth) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/core/MultiListenerServerIODispatch.java:71]: Done

**Review by sajinieKavindya — APPROVED** [review]

**Pull request merged**